Kafka-4351: Topic regex behavioral change with MirrorMaker new consumer

huxihx opened the pull request:
Support CSV values in the regexp for MirrorMaker new consumer as OldConsumer does.

huxihx committed:
Fix kafka-4351: Topic regex behavioral change with MirrorMaker new consumer
Author: huxi_2b@hotmail.com

Support CSV values in the regexp for MirrorMaker new consumer as OldConsumer does.

huxihx commented:
@joestein please help review this. Thanks.

ijuma commented:
Thanks for the PR. A couple of things:
1. This could break existing users, so we need to decide if we're OK with that. Probably worth starting a mailing list thread to get wider input. If we go ahead, we should add a note to the upgrade notes so that users are aware of the change.
2. There should be tests to verify the change.

huxihx commented:
@ijuma, totally agreed on that we need to test the change. As for the doc update thing, I think we might not need to do this since right now the document already claims:

> For convenience we allow the use of ',' instead of '|' to specify a list of topics.

And we know this is exactly what this fix is about. Without the change, only old consumer supports CSV-typed regular expression because of the fact that TopicFilter will convert them internally. But the new consumer is not able to do this. Does it make sense?

ijuma commented:
That's a fair point, if we already document it like that, then it's just a bug fix. Let's add a note to the upgrade notes still though (`upgrade.html`).

huxihx commented:
@ijuma Please guide what I should do next. I am a freshman to the process. Thanks.

vahidhashemian commented on core/src/main/scala/kafka/tools/MirrorMaker.scala:
```diff
@@ -561,7 +561,12 @@ object MirrorMaker extends Logging with KafkaMetricsGroup {
+            .replace(" ", "")
+            .replaceAll("""^["']+""","")
+            .replaceAll("""["']+$""","") // property files may bring quotes
+          consumer.subscribe(Pattern.compile(revisedWhitelist), consumerRebalanceListener)
```
What if we use something like this

```
          val whitelist = Whitelist(whitelistOpt.get)
          consumer.subscribe(Pattern.compile(whitelist.regex), consumerRebalanceListener)
```

so that regex manipulation is hardcoded in one place?

huxihx replied:
@vahidhashemian Agreed on your suggestion. And if we employs Whitelist here,  we need to additionally do a minor change for the exception capturing. Here is the analysis:

Before initializing Whitelist, the construction method for TopicFilter will be executed, which invokes Pattern.compile(regex). If an 'invalid' regex is offered, TopicFilter captures it and throws **RuntimeException**(not PatternSyntaxException), which means, later, the init method for MirrorMakerNewConsumer fails to capture this exception since it stares at PatternSyntaxException (although the impact is ignorable since the enclosing run method captures throwable)

Does it make sense? If you agree with the analysis above, I will make another commit following your suggestion and have this PR ready to be reviewed again. Is it okay to you?

vahidhashemian replied:
Thanks for explaining. It makes sense to me.
On a separate note, do you think it makes sense to add some unit test for this change?

huxihx replied:
@vahidhashemian Testing is definitely important but as for this case, I am afriad what we need to test  is whether Whitelist behaves as expected, which has already been tested in TopicFilterTest. Besides, I made a new commit following your suggestion, please have a review. Thanks.

vahidhashemian replied:
Looks like you committed more than needed in the last try?
On the unit test note, I'm wondering if it makes sense to add (at least) one unit test that verifies the issue raised in the JIRA would not occur with this patch?

huxihx replied:
@vahidhashemian  already committed a new one. And as for the unit test thing, which file do you think should be better for the test case to be added in?  TopicFilterTest or MirrorMakerTest?

vahidhashemian replied:
For the unit test I'd add it to MirrorMakerTest.

huxi-2b committed:
Merge branch 'kafka-4351_Regex_behavior_change_for_new_consumer' of https://github.com/amethystic/kafka into kafka-4351_Regex_behavior_change_for_new_consumer

huxi-2b committed:
kafka 4351: Topic regex behavioral change with MirrorMaker new consumer
author: huxi_2b@hotmail.com

Support CSV values in the regexp for MirrorMaker new consumer as
OldConsumer does.

huxi-2b committed:
kafka-4351: Topic regex behavioral change with MirrorMaker new consumer
Author: huxi_2b@hotmail.com

Support CSV values in the regexp for MirrorMaker new consumer as
OldConsumer does.

huxi-2b committed:
Merge remote-tracking branch 'upstream/trunk' into kafka-4351_Regex_behavior_change_for_new_consumer

huxihx committed:
Merge remote-tracking branch 'upstream/trunk' into kafka-4351_Regex_behavior_change_for_new_consumer

huxihx committed:
Merge branch 'kafka-4351_Regex_behavior_change_for_new_consumer' of https://github.com/amethystic/kafka into kafka-4351_Regex_behavior_change_for_new_consumer

huxihx commented:
@vahidhashemian @ijuma any feedback for this PR? Thanks.

huxihx commented:
@joestein Could you help review this PR? Thanks.

vahidhashemian commented:
@amethystic The change you made to fix the issue looks good to me. But @ijuma's concerns above should still be addressed in my opinion.

huxihx commented:
@vahidhashemian Thanks for your feedback. As for @ijuma 's suggestion, seems that he mentions of upgrade.html, so do you know how to add a update not for this PR? How should I do then ?

ijuma commented:
`upgrade.html` is in `docs` folder. Not sure what you mean by "a update not for this PR".

huxihx commented:
@ijuma Thanks. So can this PR be closed now?

ijuma commented:
@amethystic no, as it has not been merged yet. To clarify: before we merge it, we need:
1. Tests
2. Add a note to `docs/upgrade.html` mentioning this change

huxihx committed:
Merge remote-tracking branch 'upstream/trunk' into kafka-4351_Regex_behavior_change_for_new_consumer

huxihx committed:
Update docs/upgrade.html for Kafka-4351. Add statement on 'regular expression is supported for the --whitelist option of the  new consumer in MirrorMaker'

huxihx commented:
@ijuma Thanks for the advice. Already updated the upgrade note. Besides, I did a test in my local environment, which was passed. Both single topic and regular expressions are supported now for the new consumer of MirrorMaker. Do you think we need to add some unit test cases although it is not an easy work to do this?

huxi-2b committed:
Merge remote-tracking branch 'upstream/trunk' into kafka-4351_Regex_behavior_change_for_new_consumer

huxihx commented:
@ijuma could you please address the comments above? Thanks.

ijuma commented:
@amethystic, yes it would be good to add at least one test.

huxihx commented:
@ijuma Sorry, but do you mean the unit test?

huxihx committed:
Merge remote-tracking branch 'upstream/trunk' into kafka-4351_Regex_behavior_change_for_new_consumer

huxihx commented:
@ijuma Seems no MirrorMaker unit test is actually offered, although there is a MirrorMakerTest. Most of the stuffs in the MirrorMaker is private, it's a little hard to test them outside it. Could you give some advice?

huxihx committed:
Merge remote-tracking branch 'upstream/trunk' into kafka-4351_Regex_behavior_change_for_new_consumer

huxihx committed:
Merge remote-tracking branch 'upstream/trunk' into kafka-4351_Regex_behavior_change_for_new_consumer

huxihx committed:
add unit-test case for Kafka4351: Topic regex behavioral change with MirrorMaker new consumer
@author: huxi2b@gmail.com

Add a test case for MirrorMaker to verify it supports regex.

huxihx commented:
@ijuma  added a test case for this PR. Please have a review. Thanks.

huxihx committed:
remove useless imports and fix the regex string for Kafka4351: Topic regex behavioral change with MirrorMaker new consumer
@author: huxi2b@gmail.com

ijuma commented on core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala:
```diff
@@ -17,12 +17,82 @@
-import org.junit.Test
+import org.junit.{After, Before, Test}
+
+class MirrorMakerTest extends ZooKeeperTestHarness {
```
I suggest creating a new test under `integration/kafka/tools/MirrorMaker.scala` instead of reusing the existing unit test to avoid making the existing unit tests slower due to initialisation that they don't need.

Also, it seems like it may be better to inherit from `KafkaServerTestHarness` than `ZooKeeperTestHarness`?

ijuma commented on core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala:
```diff
@@ -17,12 +17,82 @@
+    val props = new Properties()
+    props.put("bootstrap.servers", brokerList)
+    props.put("key.serializer", "org.apache.kafka.common.serialization.ByteArraySerializer")
+    props.put("value.serializer", "org.apache.kafka.common.serialization.ByteArraySerializer")
```
Instead of setting the serializers like this, you can just pass them to the `KafkaProducer` constructor.

ijuma commented on core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala:
```diff
@@ -17,12 +17,82 @@
+    props.put("value.serializer", "org.apache.kafka.common.serialization.ByteArraySerializer")
+    val producer1 = new KafkaProducer[Array[Byte], Array[Byte]](props)
+    producer1.send(new ProducerRecord[Array[Byte], Array[Byte]](topic, msg.getBytes()))
+    producer1.flush() // Explicitly invoke flush method to make effect immediately
```
`close` should have the same effect as `flush`. Does the test fail if you remove `flush`?

ijuma commented on core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala:
```diff
@@ -17,12 +17,82 @@
+    props.put("key.serializer", "org.apache.kafka.common.serialization.ByteArraySerializer")
+    props.put("value.serializer", "org.apache.kafka.common.serialization.ByteArraySerializer")
+    val producer1 = new KafkaProducer[Array[Byte], Array[Byte]](props)
+    producer1.send(new ProducerRecord[Array[Byte], Array[Byte]](topic, msg.getBytes()))
```
You should probably call `get` on the returned `Future` to ensure that we get a useful error if there is a failure. Also, you can omit `[Array[Byte], Array[Byte]]` in both cases above.

ijuma commented on core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala:
```diff
@@ -17,12 +17,82 @@
+    val producer1 = new KafkaProducer[Array[Byte], Array[Byte]](props)
+    producer1.send(new ProducerRecord[Array[Byte], Array[Byte]](topic, msg.getBytes()))
+    producer1.flush() // Explicitly invoke flush method to make effect immediately
+    producer1.close() // Close the producer
```
This comment is redundant.

ijuma commented on core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala:
```diff
@@ -17,12 +17,82 @@
+    producer1.close() // Close the producer
+
+    // Create a MirrorMaker consumer
+    val config = new util.HashMap[String, AnyRef]
```
Is there a reason why we use a `Properties` instance for the producer and a `HashMap` for the consumer? It would be good to be consistent (either choice is fine).

ijuma commented on core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala:
```diff
@@ -17,12 +17,82 @@
+
+    // Create a MirrorMaker consumer
+    val config = new util.HashMap[String, AnyRef]
+    config.put(ConsumerConfig.GROUP_ID_CONFIG, "test-gropu")
```
Typo in group name.

ijuma commented on core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala:
```diff
@@ -17,12 +17,82 @@
+    config.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, brokerList)
+    config.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest")
+    config.put("key.deserializer", "org.apache.kafka.common.serialization.ByteArrayDeserializer")
+    config.put("value.deserializer", "org.apache.kafka.common.serialization.ByteArrayDeserializer")
```
Again, you can pass the deserializers via the constructor.

ijuma commented on core/src/test/scala/unit/kafka/tools/MirrorMakerTest.scala:
```diff
@@ -17,12 +17,82 @@
+    val consumer = new KafkaConsumer[Array[Byte], Array[Byte]](config)
+    MirrorMaker.createMirrorMakerProducer(brokerList)
+
+    val whitelist = Some("new.*|another_topic|foo" ) // Test a regular expression
```
I thought we wanted to test multiple regular expressions separated by commas?

huxihx replied:
I replaced all "|" with commas. Is that what you expect?

ijuma replied:
Yeah, that should fail without your fix, I think. If we wanted to make the test even better, we'd produce to `another_topic` and `yet_another_topic` and check that the consumer only consumes from the two that match.

But the test as it is verifies the fix, so I won't block the merge if you don't feel like extending the test. :)

ijuma commented on core/src/main/scala/kafka/tools/MirrorMaker.scala:
```diff
@@ -310,6 +310,22 @@ object MirrorMaker extends Logging with KafkaMetricsGroup {
   }
 
+  // Only for testing
+  private[kafka] def createMirrorMakerProducer(brokerList: String): Unit = {
```
Why is this method here? Can it not be in the test?

ijuma commented on core/src/main/scala/kafka/tools/MirrorMaker.scala:
```diff
@@ -561,9 +578,10 @@ object MirrorMaker extends Logging with KafkaMetricsGroup {
       val consumerRebalanceListener = new InternalRebalanceListenerForNewConsumer(this, customRebalanceListener)
       if (whitelistOpt.isDefined) {
```
Instead of `whitelistOpt.isDefined`, we can write it as:

```scala
whitelistOpt.foreach { whitelist =>
  try {
    consumer.subscribe(Pattern.compile(Whitelist(whitelist).regex), consumerRebalanceListener)
    ...
  }
}
```

ijuma replied:
What are your thoughts on this comment?

ijuma reviewed (changes requested):
Sorry for the delay. Thanks for adding the test, very useful. I left a few suggestions and questions.

huxihx committed:
Merge remote-tracking branch 'upstream/trunk' into kafka-4351_Regex_behavior_change_for_new_consumer

huxihx committed:
kafka--4351: Topic regex behavioral change with MirrorMaker new consumer
1. Move unit test case under integration/kafka/tools
2. Remove test code in MirrorMaker

huxihx commented:
@ijuma already addressed most of your comments. Please have a review. Thanks

ijuma commented on core/src/main/scala/kafka/tools/MirrorMaker.scala:
```diff
@@ -574,7 +574,8 @@ object MirrorMaker extends Logging with KafkaMetricsGroup {
 
-  private class MirrorMakerNewConsumer(consumer: Consumer[Array[Byte], Array[Byte]],
+  // Only for testing
+  private[kafka] class MirrorMakerNewConsumer(consumer: Consumer[Array[Byte], Array[Byte]],
```
Should this be `private[tools]`?

ijuma commented on core/src/main/scala/kafka/tools/MirrorMaker.scala:
```diff
@@ -686,7 +688,7 @@ object MirrorMaker extends Logging with KafkaMetricsGroup {
   }
 
-  private class MirrorMakerProducer(val producerProps: Properties) {
+  private[kafka] class MirrorMakerProducer(val producerProps: Properties) {
```
Should this be `private[tools]`?

ijuma commented on core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala:
```diff
@@ -0,0 +1,69 @@
+  @After
+  override def tearDown() {
+    super.tearDown()
+  }
```
We don't need `tearDown` and `setUp` defined here if they just call `super`.

ijuma commented on core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala:
```diff
@@ -0,0 +1,69 @@
+    mirrorMakerConsumer.init()
+    try {
+      val data = mirrorMakerConsumer.receive()
+      println(new String(data.value))
```
Probably left unintentionally?

ijuma commented on core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala:
```diff
@@ -0,0 +1,69 @@
+    try {
+      val data = mirrorMakerConsumer.receive()
+      println(new String(data.value))
+      assertTrue(s"MirrorMaker consumer should get the correct topic: $topic", data.topic.equals(topic))
```
In Scala, `data.topic == topic` is more idiomatic. Same for the next line.

ijuma commented on core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala:
```diff
@@ -0,0 +1,69 @@
+      assertTrue(s"MirrorMaker consumer should get the correct topic: $topic", data.topic.equals(topic))
+      assertTrue("MirrorMaker consumer should read the correct message.", new String(data.value).equals(msg))
+    } catch {
+      case e: RuntimeException =>
```
No need to catch and rethrow, just let the original exception bubble up.

ijuma commented on core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala:
```diff
@@ -0,0 +1,69 @@
+    // Create a test producer to delivery a message
+    val producerProps = new Properties()
+    producerProps.put("bootstrap.servers", brokerList)
+    producerProps.setProperty(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer")
```
You can set the class itself instead of the name, I think. e.g. `classOf[ByteArraySerializer]`. Same in the line below.

ijuma commented on core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala:
```diff
@@ -0,0 +1,69 @@
+import org.junit.Assert.assertTrue
+import org.junit.{After, Before, Test}
+
+class TestMirrorMaker extends KafkaServerTestHarness {
```
I thought `MirrrorMakerTest` was fine, why did you rename it?

huxihx replied:
The compiler will complain "MirrorMakerTest is already defined as class MirrorMakerTest".   That's why I changed the name. Do you think there is another way I could avoid this error?

ijuma replied:
Ah, that makes sense because they both have the same package (it's just the source folder that differs). In that case, I suggest `MirrorMakerIntegrationTest` for this one so the difference is obvious.

ijuma commented on core/src/test/scala/integration/kafka/tools/TestMirrorMaker.scala:
```diff
@@ -0,0 +1,69 @@
+    producerProps.setProperty(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer")
+    val producer = new MirrorMakerProducer(producerProps)
+    MirrorMaker.producer = producer
+    MirrorMaker.producer.send(new ProducerRecord(topic, msg.getBytes()))
```
As I said before, it would be good to call `get` on the result of `send` to ensure exceptions get bubbled up.

huxihx replied:
The producer here is not of type 'KafkaProducer' but 'MirrorMakerProducer'. It's send method is already KafkaProducer.send().get() by default if not explicitly specifying 'async'.  Does it make sense?

ijuma replied:
Yes, thanks for clarifying.

ijuma replied:
Hmm, the default seems to be `async`?

```scala
val sync = producerProps.getProperty("producer.type", "async").equals("sync")
```

huxihx replied:
Sorry, my mistake. Do you think it's okay to add props.put("producer.type", "sync") before MirrorMakerProducer?

ijuma replied:
Sounds good to me.

ijuma reviewed (changes requested):
Thanks, I think we're close. Left a few more comments.

huxi-2b committed:
Merge branch 'kafka-4351_Regex_behavior_change_for_new_consumer' of https://github.com/amethystic/kafka into kafka-4351_Regex_behavior_change_for_new_consumer

huxi-2b committed:
kafka-4351: Topic regex behavioral change with MirrorMaker new consumer
1. narrow down the visibility for some method and class
2. modify test case

huxi-2b committed:
Merge remote-tracking branch 'upstream/trunk' into kafka-4351_Regex_behavior_change_for_new_consumer

huxihx commented:
@ijuma Thanks for your great patience. Please review it again after addressing your comments. Thanks.

ijuma commented on core/src/test/scala/integration/kafka/tools/MirrorMakerIntegrationTest.scala:
```diff
@@ -0,0 +1,68 @@
+          assertTrue("MirrorMaker consumer should read the correct message.", new String(data.value) == msg)
+          return
+        } catch {
+         case _: ConsumerTimeoutException => // swallow it
```
Hmm, this doesn't seem right. Why do we swallow this exception?

huxihx replied:
`val mirrorMakerConsumer = new MirrorMakerNewConsumer(consumer, None, whitelist)

mirrorMakerConsumer.init()

 try {
      TestUtils.waitUntilTrue(() => {
        try {
          val data = mirrorMakerConsumer.receive()
          data.topic == topic && new String(data.value) == msg
        } catch {
          case _: ConsumerTimeoutException => false
        }
      }, "MirrorMaker consumer should be able to subscribe the correct topic and read the correct message.")
    } finally {
      consumer.close()
}`
Is that okay to you then? Thanks.

ijuma commented on core/src/test/scala/integration/kafka/tools/MirrorMakerIntegrationTest.scala:
```diff
@@ -0,0 +1,68 @@
+    mirrorMakerConsumer.init()
+    try {
+      val maxTryCount = 3 // it might need to call multiple poll calls to retrieve the message
+      for (_ <- 0 until maxTryCount) {
```
We should use `TestUtils.waitUntilTrue` instead of this.

ijuma commented on core/src/test/scala/integration/kafka/tools/MirrorMakerIntegrationTest.scala:
```diff
@@ -0,0 +1,68 @@
+    MirrorMaker.producer.close()
+
+    servers foreach { server =>
+      println(server.zkUtils.getAllTopics().mkString(","))
```
We shouldn't have this. Unintentional?

huxihx replied:
I always forget to remove them before committing. Sorry, my mistake.

ijuma commented on core/src/test/scala/integration/kafka/tools/MirrorMakerIntegrationTest.scala:
```diff
@@ -0,0 +1,68 @@
+    producerProps.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, brokerList)
+    producerProps.put("producer.type", "sync")
+    producerProps.setProperty(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer")
+    producerProps.setProperty(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, "org.apache.kafka.common.serialization.ByteArraySerializer")
```
Did it not work to use `classOf[ByteArraySerializer]` instead of hardcoding the String?

huxihx replied:
Seems MirrorMakerProducer is not a parameterized class which cannot use type parameters during construction.

ijuma replied:
That's not what I was suggesting. That's OK, I'll do it before merging.

ijuma reviewed (changes requested):
Thanks for the updates, a few more comments.

amethystic commented on core/src/test/scala/integration/kafka/tools/MirrorMakerIntegrationTest.scala:
```diff
@@ -0,0 +1,68 @@
+          assertTrue("MirrorMaker consumer should read the correct message.", new String(data.value) == msg)
+          return
+        } catch {
+         case _: ConsumerTimeoutException => // swallow it
```
`val mirrorMakerConsumer = new MirrorMakerNewConsumer(consumer, None, whitelist)
    mirrorMakerConsumer.init()
    try {
      TestUtils.waitUntilTrue(() => {
        try {
          val data = mirrorMakerConsumer.receive()
          data.topic == topic && new String(data.value) == msg
        } catch {
          case _: ConsumerTimeoutException => false// swallow it
        }
      }, "MirrorMaker consumer should be able to subscribe the correct topic and read the correct message.")
    } finally {
      consumer.close()
    }`
Is that okay to you? Thanks.

huxi-2b committed:
kafka-4351: Topic regex behavioral change with MirrorMaker new consumer
fix MirrorMakerIntegrationTest.testRegularExpressionTopic code as per
Ijuma's comments.

ijuma commented:
Thanks for the updates, LGTM with a few minor changes that I will do before merging to trunk: moved upgrade notes to the right place and reworded them a little, added license comment to test, minor style improvements.

huxihx commented:
Thanks for your great patience. I do learn a lot from this whole review cycle.
